### ISSUE TITLE
Advanced Android support

### ISSUE THREAD

**Issue by shanshin** [Kotlin/kotlinx-kover#18]
It is necessary to improve the plugin API for convenient work with Android applications:
- support for Android build types
- support Android flavors
- consider creating merge reports for each Build Type

**Comment by pvegh** [comment]
Cannot use this until then. It's not realistic for us currently.
<===========--> 86% EXECUTING [25m 16s]

**Comment by shanshin** [comment]
@pvegh, please, can you clarify how long it takes to build a project without applying a Kover, how many tests and classes are in the project.
Perhaps you have suggestions on what settings can be added to the plugin to reduce code instrumentation and speed up your build.

**Comment by pvegh** [comment]
Those details are not important, the problem is that ./gradlew koverHtmlReport is not flavour specific.
Currently this one Gradle task builds the project in every possible variant, which means debug + release build types for all of paid, free, foo flavours.
Meaning it builds 6 times instead of just one (whatever is needed e.g. freeDebug for local or freeRelease for CI)

**Comment by SpertsyanKM** [comment]
@pvegh I've solved this issue by simply adding this configuration to `android.testOptions.unitTests.all` section in `build.gradle`.
```
            kover {
                disabled = (name == "testReleaseUnitTest")
            }
```
May be it will help you?

**Comment by pvegh** [comment]
Sorry I didn't have time to test this. I'll try it next week. Thanks!

**Comment by pvegh** [comment]
I see now that we removed kover from the project, because it wasn't feasible until variants are correctly supported. I don't think disabling variants one by one is a good practice.
If I execute ./gradlew koverHtmlReportLocalDebug then no _release_, no _integration_ etc should be executed.

**Comment by subspider** [comment]
we have the same problem can we at least have the possibility to avoid certain flavors ?

**Comment by gerardoepitacio** [comment]
@shanshin Is there a way to see the contribution guidelines for this project?, I really wanted to learn how to write a gradle plugin to be able to improve this tool, I would really appreciate a guideline over that.

**Comment by NeoMindStd** [comment]
We need report tasks support to specific buildVariant.
Do you have any plans or schedules to support?

**Comment by shanshin** [comment]
@NeoMindStd, yes, right now we are developing a new version of the plugin API, after switching to which filtering by buildVariant and/or flavours will be added.

**Comment by shanshin** [comment]
@gerardoepitacio, at the moment there are no strict documents, you can create PR with your suggestions and bug fixes.
Because the project code is not stabilized at the moment, I would not recommend creating large PR.  However, if you have ideas, you can create an issue with their description.

**Comment by peterArconsis** [comment]
I think this should be added before 1.0.0 because this way it's not feasible to use/test it on big projects, where only 1 variant takes 5-10 minutes. Building unnecessary variants takes up to an hour. We'd like to test this and provide feedback, currently we can't.

**Comment by shanshin** [comment]
@samuelneff, work is in progress now.
since this is not just a bug, but new features that were not originally laid in the plugin API, it took a redesign of the API for more convenient use in Android.

**Comment by shanshin** [comment]
@samuelneff, however, it would help us a lot to know how you build reports when there are several build types and flavors in the project.

Do you generate only one report by build type and all flavors, multiple reports at a time for different build types, or only individual reports for each build variant (build type + flavor)?

**Comment by G00fY2** [comment]
@samuelneff We have an Android project with a lot of variants (multiple flavors in different dimensions + multiple build types). We are currently using Jacoco without any issues. But even after disabling all tests tasks expect for one variant we got `attempted duplicate class definition` exceptions in our build. So I guess there are more underlying issues which have to be fixed instead of only limiting the task execution.

**Comment by samuelneff** [comment]
@G00fY2 

> We are currently using Jacoco without any issues

We actually have been running with Kover for about a year without realizing we had this issue. We always just wondered why our builds were so slow without making it a priority to look into it. It's pretty common in my experience for builds to be slower on CI than locally. 

We only discovered the problem when we made a change to our build to surface errors as a giant annotation instead of being buried in the CI console output. Once we did that we immediately saw that all errors were showing up 12 times each.

**Comment by samuelneff** [comment]
@G00fY2 Thanks for the recommendation. I appreciate it, but am not sure it applies to our (unusual?) use case.

While we only build one variant at a time, we would want to run the unit tests regardless of what variant was run. If we always ran the same variant, like `testDevDebugUnitTest` then that would require compiling that variant regardless of which one we actually wanted to build. So we would often end up compiling twice, which is not necessary in our specific case and what I'm looking to avoid.

**Comment by hansenji** [comment]
My team is in a similar space as @samuelneff. We have a few build types that are mutually exclusive and running kover on all build types takes valuable of cpu time.

**Comment by sonrohan** [comment]
+1 Support for this

**Comment by adrianlazaro8** [comment]
+1 for this advanced Android support. Thanks for your work!

**Comment by outadoc** [comment]
@G00fY2 **Thank you** for this snippet, it solves our exact needs, for both Kover and just configuring unit tests correctly overall! I'd been trying to figure this out for quite some time 🎉

**Comment by jweick** [comment]
Hi,

i followed the discussion what can be done to avoid running all the buildType/variants when executing `koverReport` and tried a variant of a mentioned configuration option

> kover {
    ~~disabled = (name == "testReleaseUnitTest")~~
    enabled = name == "testDebugUnitTest"
}

to only run it on the `testDebugUnitTest` task.

That worked only partly, as something like `testReleaseUnitTest` was indeed not executed (log said it was SKIPPED), but related tasks like `bundleReleaseClassesToCompileJar` or `compileReleaseUnitTestKotlin` where executed.
And that should not happen, as the job just builds a debug-build.

Any ideas for a workaround until kover supports buildTypes/variants?

regards,
Jerg

**Comment by shanshin** [comment]
Resolved in `0.7.0-Alpha`

**Comment by jtietema** [comment]
I am having the same issue as @hamada147, also getting the duplicate class errors using the Jacoco engine. I am running the freshly released Kover `0.8.0`.
